### PR TITLE
Update the docs for black

### LINKED ISSUE #13948: Update the documentation for `black`
Per [APE 20](https://github.com/astropy/astropy-APEs/blob/main/APE20.rst) and noted by in https://github.com/astropy/astropy/pull/13253#issuecomment-1297595911 the  [coding style](https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions) documentation needs to be updated to reflect the adoption of black. Moreover the documentation needs to be searched to remove any references to any other coding syles.

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This PR updates the developer guide to detail the usage of `black`.

Resolves #13948

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: docs/development/codeguide.rst [hunk 1]
```diff
@@ -135,8 +135,17 @@ Coding Style/Conventions
   <https://www.python.org/dev/peps/pep-0008/>`_. In particular, this includes
   using only 4 spaces for indentation, and never tabs.
 
+  * ``astropy`` itself enforces this style guide using the
+    `black <https://black.readthedocs.io/en/stable/>`_ code formatter.
+
+  * We recognize that sometimes ``black`` will autoformat things in undesirable
+    ways e.g. matricies.  In the cases that ``black`` produces undesirable code
```
Comment (pllim) [inline docs/development/codeguide.rst:142]:
```suggestion
    ways, e.g., matrices.  In the cases that ``black`` produces undesirable code
```

File: docs/development/codeguide.rst [hunk 2]
```diff
@@ -135,8 +135,17 @@ Coding Style/Conventions
...
+    formatting, one can wrap code the code in ``# fmt: off`` and ``# fmt: on``
```
Comment (pllim) [inline docs/development/codeguide.rst:143]: So they have to be in pairs? What if I only want to disable the one line, is there an equivalent of `# noqa` for `flake8`?

File: docs/development/codeguide.rst [hunk 3]
```diff
@@ -135,8 +135,17 @@ Coding Style/Conventions
...
+    to disable ``black`` formatting. This should be done sparingly, and only
+    when ``black`` produces undesirable formatting.
+
 * Our testing infrastructure currently enforces a subset of the PEP8 style
-  guide, and some sub-packages enforce stronger styling checks such as using
+  guide. In addition these checks also enforce
```
Comment (pllim) [inline docs/development/codeguide.rst:148]:
```suggestion
  guide. In addition, these checks also enforce
```

File: docs/development/codeguide.rst [hunk 4]
```diff
@@ -135,8 +135,17 @@ Coding Style/Conventions
   <https://www.python.org/dev/peps/pep-0008/>`_. In particular, this includes
   using only 4 spaces for indentation, and never tabs.
 
+  * ``astropy`` itself enforces this style guide using the
```
Comment (pllim) [inline docs/development/codeguide.rst:138]: Maybe we should mention the PR or hash when this was "activated" in the branch? Or maybe that doesn't matter?

Didn't @eteq suggest some special tag?

File: docs/development/codeguide.rst (hunk 2 above)
Comment (WilliamJamieson) [reply docs/development/codeguide.rst:143]: Possibly, here is what the docs say: https://black.readthedocs.io/en/stable/faq.html#why-does-my-linter-or-typechecker-complain-after-i-format-my-code

Comment (pllim) [reply docs/development/codeguide.rst:143]: I don't think that answers my question. Is there like a `# noblack` or `# fmt: skip` if I only want to opt-out of black for that one line?

Comment (WilliamJamieson) [reply docs/development/codeguide.rst:143]: Oh, I see what you are asking. Yes `# fmt: skip` is the official way to do it per https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#code-style.

Comment (pllim) [reply docs/development/codeguide.rst:143]: Great! Perhaps that should be mentioned here too? That might be very common use case in this repo... 💭

**Review by pllim — APPROVED** [review]
LGTM but I would like to hear from @eteq whether we need special tags or not.

**Comment by eerovaher** [comment]
In some cases it is possible to influence `black` formatting with a [magic trailing comma](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma). For example, small matrices can be written down in the following way without using any `black` switches:
```python
matrix = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1],
]
```
I think this can be useful enough that it should be pointed out in our documentation.

**Comment by WilliamJamieson** [comment]
> In some cases it is possible to influence `black` formatting with a [magic trailing comma](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma). For example, small matrices can be written down in the following way without using any `black` switches:
> 
> ```python
> matrix = [
>     [1, 0, 0],
>     [0, 1, 0],
>     [0, 0, 1],
> ]
> ```
> 
> I think this can be useful enough that it should be pointed out in our documentation.

I think this type of information warrants a new section in the docs. Though I don't know the exact details of what should be in it, or where we should include it.

**Comment by pllim** [comment]
The trailing comma really annoys me.

**Comment by WilliamJamieson** [comment]
> The trailing comma really annoys me.

I kinda like the trailing comma, it often lines up with the commas on all the lines above.

**Comment by pllim** [comment]
Trailing comma is syntax error in some other languages. I am afraid this will teach me bad habit.

**Comment by astrofrog** [comment]
I don't mind the trailing comma,

**Comment by dhomeier** [comment]
I find it disturbing, too, but that's just what black does to sequences, right?
https://github.com/astropy/astropy/blob/6400c89e58add0e8b96e00f26dcd48b877c2acc9/astropy/io/ascii/rst.py#L8-L13
But if it works, important enough to know – these possibly not easy to find tricks should definitely be in the docs.

**Review by astrofrog — APPROVED** [review]
Looks good to me! I'll leave this a little longer but then we can merge even during the Black transition period as it only touches the docs.

**Comment by pllim** [comment]
@astrofrog , ok to merge now?

**Pull request merged**